### PR TITLE
feat: Add protect & unprotect workunit to WU context menu

### PR DESCRIPTION
**Pull request by dehilsterlexis**
This will require the adding of two functions to the Workunit class in the comms code in the Visualization repository: https://github.com/hpcc-systems/Visualization/blob/trunk/packages/comms/src/ecl/workunit.ts

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Have you opened a PR or Ticket for the Viz repo?

Can you remove the submenu and only show "Protect" when WU is not protected and show "Unprotect" when the WU is protected?

**Comment by dehilsterlexis** [comment]
I did create a ticket (not an issue) for this.

**Comment by dehilsterlexis** [comment]
@GordonSmith will make the changes you requested.

**Comment by dehilsterlexis** [comment]
@GordonSmith come to think of it, I don't believe you can dynamically change the menu. You can only gray things out. So I suggest we gray the appropriate one out.

**Comment by GordonSmith** [comment]
It is a little bit tricky - but if you search the code base for:
* hpccPlatform.isAllWorkunits
* hpccPlatform.isMyWorkunits

You can see how I worked around the issue.

**Comment by dehilsterlexis** [comment]
@GordonSmith the setContext command is a global command and works in the case of isAllWorkunits and isMyWorkunits because they are globals that pertain to all tree nodes. The protect only applies to a single node so it can't be used in the same way. It will have to be grayed out using viewItem because viewItem is specific to each tree node.

**Comment by GordonSmith** [comment]
Sounds plausible.

**Comment by GordonSmith** [comment]
@dehilsterlexis you need to 
1. Bump the version of @hpcc-js/comms in the package.json
2. Run `npm install`
3. Commit package.json + package-lock.json

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: src/ecl/eclWatchTree.ts [hunk 1]
```diff
@@ -93,6 +93,16 @@ export class ECLWatchTree extends Tree {
             wuNode.delete();
         });
 
+        vscode.commands.registerCommand("hpccPlatform.protectWU", (wuNode: ECLWUNode) => {
+            wuNode.protect();
+            this.refresh();
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:98]: Not sure this refresh is needed?  If it is, it should wait for the "protect" to complete?

File: src/ecl/eclWatchTree.ts [hunk 2]
```diff
@@ -93,6 +93,16 @@ export class ECLWatchTree extends Tree {
...
+        });
+
+        vscode.commands.registerCommand("hpccPlatform.unprotectWU", (wuNode: ECLWUNode) => {
+            wuNode.unprotect();
+            this.refresh();
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:103]: Not sure this refresh is needed?  If it is, it should wait for the "unprotect" to complete?

File: src/ecl/eclWatchTree.ts [hunk 3]
```diff
@@ -400,6 +410,14 @@ export class ECLWUNode extends Item<ECLWatchTree> {
         this._wu.abort().then(() => this._tree._onDidChangeTreeData.fire(this));
     }
 
+    protect() {
+        this._wu.protect();
+    }
+
+    unprotect() {
+        this._wu.unprotect();
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:418]: Should refresh the tree?

File: src/ecl/eclWatchTree.ts [hunk 4]
```diff
@@ -400,6 +410,14 @@ export class ECLWUNode extends Item<ECLWatchTree> {
         this._wu.abort().then(() => this._tree._onDidChangeTreeData.fire(this));
     }
 
+    protect() {
+        this._wu.protect();
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:414]: Should refresh the tree?

File: src/ecl/eclWatchTree.ts [hunk 5]
```diff
@@ -421,7 +439,8 @@ export class ECLWUNode extends Item<ECLWatchTree> {
     }
 
     contextValue(): string {
-        return this._wu.isComplete() ? "ECLWUNodeComplete" : "ECLWUNode";
+        const prot = this._wu.Protected ? "Protected" : "Unprotected";
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:442]: Change to `ECLWUNodeProtected` and `ECLWUNode` respectivly.

File: src/ecl/eclWatchTree.ts [hunk 6]
```diff
@@ -421,7 +439,8 @@ export class ECLWUNode extends Item<ECLWatchTree> {
...
+        return this._wu.isComplete() ? `ECLWUNodeComplete${prot}` : `ECLWUNode${prot}`;
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:443]: Change to 
```ts
return this._wu.isComplete() ? `ECLWUNodeComplete,${prot}` : `ECLWUNode,${prot}`;
```
respectively

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: src/ecl/eclWatchTree.ts (hunk 1 above)
Comment (dehilsterlexis) [reply src/ecl/eclWatchTree.ts:98]: Yes, you need it or the menu doesn't update.

File: src/ecl/eclWatchTree.ts (hunk 2 above)
Comment (dehilsterlexis) [reply src/ecl/eclWatchTree.ts:103]: Yes, you need it or the menu doesn't update.

File: src/ecl/eclWatchTree.ts (hunk 4 above)
Comment (dehilsterlexis) [reply src/ecl/eclWatchTree.ts:414]: No, that would call it twice.

File: src/ecl/eclWatchTree.ts (hunk 3 above)
Comment (dehilsterlexis) [reply src/ecl/eclWatchTree.ts:418]: No, that would call it twice.

File: src/ecl/eclWatchTree.ts (hunk 6 above)
Comment (dehilsterlexis) [reply src/ecl/eclWatchTree.ts:443]: Yes. I needed to simplify here after testing.

File: src/ecl/eclWatchTree.ts (hunk 1 above)
Comment (GordonSmith) [reply src/ecl/eclWatchTree.ts:98]: Remove `this.refresh` here.

File: src/ecl/eclWatchTree.ts (hunk 2 above)
Comment (GordonSmith) [reply src/ecl/eclWatchTree.ts:103]: Remove `this.refresh` here.

File: src/ecl/eclWatchTree.ts (hunk 4 above)
Comment (GordonSmith) [reply src/ecl/eclWatchTree.ts:414]: Change to:          
```
this._wu.protect().then(() => this._tree.refresh());
```

File: src/ecl/eclWatchTree.ts (hunk 3 above)
Comment (GordonSmith) [reply src/ecl/eclWatchTree.ts:418]: Change to:  
```
this._wu.unprotect().then(() => this._tree.refresh());
```

File: src/ecl/eclWatchTree.ts (hunk 5 above)
Comment (GordonSmith) [reply src/ecl/eclWatchTree.ts:442]: Bump

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Also update `ECLWUNode.getLabel` to
```
    getLabel(): string {
        let primary = this._wu.Wuid;
        const extras: string[] = [];
        if (!this._wu.isComplete() || this._wu.isDeleted()) extras.push(this._wu.State);
        if (!this._tree._myWorkunits && this._wu.Owner) extras.push(this._wu.Owner);
        if (this._wu.Jobname) {
            primary = this._wu.Jobname;
            extras.push(this._wu.Wuid);
        }
        if (this._wu.Protected) {
            extras.push("Protected");
        }
        return extras.length ? `${primary} (${extras.join(", ")})` : primary;
    }
```
And / or update the icon to show the lock symbol?

File: src/ecl/eclWatchTree.ts (hunk 5 above)
Comment (GordonSmith) [reply src/ecl/eclWatchTree.ts:440]: The reasons for this change:
1. Consistency
2. Avoid a context clash with other extensions (not very likely, but...)

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by dehilsterlexis** [comment]
@GordonSmith for review. The caret in the when clause does not seem to work in the when clauses so `ECLWUNodeUnprotected` was used.

**Review by GordonSmith — APPROVED** [review]

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Comment by dehilsterlexis** [comment]
@GordonSmith this is most likely failing because of a version problem. Any suggestions?

**Comment by GordonSmith** [comment]
You need to:
```
npm install
```
And then commit both the package-lock.json and the package.json.

Did you test its working?

**Comment by GordonSmith** [comment]
PR tests failing...

**Comment by dehilsterlexis** [comment]
@GordonSmith yes I know. I can't seem to get rid of some files in the PR that should not be there.

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: package.json [hunk 7]
```diff
@@ -401,6 +401,18 @@
         "title": "%Delete Workunit%",
         "enablement": "viewItem =~ /ECLWUNodeComplete/"
       },
+      {
+        "command": "hpccPlatform.protectWU",
+        "category": "ECL",
+        "title": "%Protect Workunit%",
+        "enablement": "viewItem =~ /ECLWUNodeComplete/"
```
Comment (GordonSmith) [inline package.json:408]: I think its ok to protect a WU while its running?

File: package.json [hunk 8]
```diff
@@ -401,6 +401,18 @@
...
+      },
+      {
+        "command": "hpccPlatform.unprotectWU",
+        "category": "ECL",
+        "title": "%Unprotect Workunit%",
+        "enablement": "viewItem =~ /ECLWUNodeComplete/"
```
Comment (GordonSmith) [inline package.json:414]: I think its ok to unprotect a WU while its running?

File: src/ecl/eclWatchTree.ts [hunk 9]
```diff
@@ -303,11 +311,7 @@ export class ECLResultNode extends Item<ECLWatchTree> {
     }
 
     getLabel(): string {
-        return this._result.Name;
-    }
-
-    getDescription(): string {
-        return this._result.Value ?? "";
+        return `${this._result.Name}:  ${this._result.Value}`;
     }
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:315]: I prefer the label + description approach.

File: src/ecl/eclWatchTree.ts [hunk 10]
```diff
@@ -371,13 +375,18 @@ export class ECLWUNode extends Item<ECLWatchTree> {
...
             extras.push(this._wu.Wuid);
         }
-        if (!this._tree._myWorkunits && this._wu.Owner) extras.push(this._wu.Owner);
-        if (!this._wu.isComplete() || this._wu.isDeleted()) extras.push(this._wu.State);
-        return extras.join(" ");
+        if (this._wu.Protected) {
+            extras.push("Protected");
+        }
+        return extras.length ? `${primary} (${extras.join(", ")})` : primary;
     }
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:390]: Wuid can end up being displayed twice (label and description) - also now its been moved to the description the parenthesis are not needed?, how about:
```ts
    getDescription(): string {
        const extras: string[] = [];
        if (this._wu.Jobname) extras.push(this._wu.Wuid);
        if (this._wu.Protected) extras.push("Protected");
        if (!this._wu.isComplete() || this._wu.isDeleted()) extras.push(this._wu.State);
        if (this._wu.Owner) extras.push(this._wu.Owner);
        if (this._wu.Cluster) extras.push(this._wu.Cluster);
        return extras.join(", ");
    }
```

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Review by GordonSmith — CHANGES_REQUESTED** [review]
WU Can be protected / unprotected while running

**Comment by GordonSmith** [comment]
@dehilsterlexis some comments and needs a rebase.

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: package.json (hunk 7 above)
Comment (GordonSmith) [reply package.json:419]: I think its ok to protect a WU while its running?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Review by GordonSmith — CHANGES_REQUESTED** [review]
You should be able to protect and unprotect a WU before its completed.

**Comment by ddehilster** [comment]
@GordonSmith now functions before completed. For review.

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**